src/wallet/rpcwallet.cpp

```diff
@@ -2117,6 +2117,77 @@ static UniValue listaccounts(const JSONRPCRequest& request)
     return ret;
 }
 
+static UniValue listsincetx(const JSONRPCRequest& request)
+{
+    CWallet* const pwallet = GetWalletForJSONRPCRequest(request);
+    if (!EnsureWalletIsAvailable(pwallet, request.fHelp)) {
+        return NullUniValue;
+    }
+
+    if (request.fHelp || request.params.size() < 1 || request.params.size() > 2)
+        throw std::runtime_error(
+            "listsincetx \"txid\" (polltime)\n"
+            "\nGet all transactions since a specific transaction [txid].\n"
+            "This call can be used as longpoll notification channel if follow-up transactions from the latest transaction-id are requested and [polltime] is greater then zero.\n"
+            "\nArguments:\n"
+            "1. \"txid\"               (string, required) The txid to list transactions since\n"
+            "1. \"polltime\"           (numeric, optional) If greater then zero, wait with a timeout of <polltime> seconds for new transactions. It will response immediately when new transactions arrive (default is 0)\n"
+            "\nExamples:\n"
+            + HelpExampleCli("listsincetx", "\"66f1e20f43780a67762fa00524e0c4ce0416339aa84e830d90e8dc636cd836e4\" 10")
+            + HelpExampleRpc("listsincetx", "\"66f1e20f43780a67762fa00524e0c4ce0416339aa84e830d90e8dc636cd836e4\", 10")
+        );
+
+    // Make sure the results are valid at least up to the most recent block
+    // the user could have gotten from another RPC command prior to now
+    pwallet->BlockUntilSyncedToCurrentChain();
+
+    // we assume that pwallet gets never deconstructed at this point
+    uint256 hash;
+    hash.SetHex(request.params[0].get_str());
+    auto it_find = pwallet->mapWallet.find(hash);
+    if (it_find == pwallet->mapWallet.end()) {
+        throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid or non-wallet transaction id");
+    }
+
+    UniValue transactions(UniValue::VARR);
+
+    // construct a function to populate the output array
+    auto addTransactionsSince = [&transactions, &pwallet](uint256 wtxid_since) {
+        LOCK2(cs_main, pwallet->cs_wallet); // we only lock at this point to allow non blocking long polling
```

Comment:
Why lock `cs_main`?

Reply:
IMO `cs_main` should always be locked before cs_wallet, especially with the usage of `ListTransactions()` and its subclass.